packages/docs/docs/docs/model/03-encode.md

```diff
@@ -1 +1,152 @@
+that, let's look into it:
+
+```ts showLineNumbers
+const model = types.string()
+const encodingResult = model.encode("foo")
+// -> ok("foo")
+```
+
+The result of the encoding process has type `Result`, a special type
+that is returned by functions that may fail with some kind of error.
+
+In the case of encoding, the value gets validated before being encoded. Since
+the process of validation may fail for invalid values, encoding can fail as well
+if you try to encode a value that is not valid.
+
+Let's look at an example:
+
+```ts showLineNumbers
+type NonNegativeNumber = types.Infer<typeof nonNegativeNumber> // -> inferred as number
+const nonNegativeNumber = types.number({ minimum: 0 })
+
+nonNegativeNumber.encode(10) // -> ok(10)
+nonNegativeNumber.encode(-1) // -> error([ assertion: "expected a number >= 0", got: -1, path: "$" ])
+```
+
+Since additional assertions described in the type options do not change the
+inferred type, it is not possible to statically check that values of the
+inferred type respect those invariants: it is the encoder's job to make sure an
+invalid value never gets encoded.
+
+In the previous example, `NonNegativeNumber` is inferred as a simple
+Typescript `number`, so the `encode` method has to check that it is a
+non-negative number. When given a negative number the encoding process fails
+with an error describing what went wrong.
+
+## A small introduction on the `Result` type
+
+Since `encode`'s return type is a `Result`, it may be useful to take a
+moment and get a feeling of how you can work with this kind of values.
+All the useful definitions can be imported from the `result` namespace of
+`@mondrian-framework/model`:
+
+```ts showLineNumbers
+import { result } from "@mondrian-framework/model"
+```
+
+A `Result<A, E>` is the return type of a function that may either fail with an
+error of type `E`, or succeed producing a value of type `A`.
+
+### Building a `Result`
+
+To build a value of type `Result` you can use the two functions `result.ok`
+and `result.fail`:
+
+```ts showLineNumbers
+const success: Result<number, string> = result.ok(10)
+const failure: Result<number, string> = result.fail("error!")
+```
+
+Consider for example a function that performs safe division, returning
+an error if the dividend is 0 instead of returning `Infinity`. It could be
+implemented using `Result`:
+
+```ts showLineNumbers
+function safeDivide(dividend: number, divisor: number): Result<number, string> {
+    return divisor === 0
+        ? result.fail("Division by 0")
+        : result.ok(dividend / divisor)
+}
+```
+
+### Working with `Result`s
+
+The great advantage provided by the result type is that it _forces_ you to deal
+with the error case, focusing your attention not only on the happy path but also
+on what could go wrong.
+
+To get a value out of a `Result` one must first check wether it is successful or
+not:
+
+```ts showLineNumbers
+function printResult(res: Result<number, string>) {
+    // highlight-start
```

Comment:
highlight-start and highlight-end should do something visually?

Reply:
Yes, they highlight the lines between `start` and `end` and hide those two comments